### ISSUE TITLE
Unable to create /etc/opt/senzing if it already exists

### ISSUE THREAD

**Issue by joshvid** [Senzing/docker-init-container#62]
**Describe the bug**
The init-container.py is unable to create /etc/opt/senzing with current versions of Python 3 (3.7)

**To Reproduce**

1. Attempt to build a wrapped container with: 

export BASE_IMAGE="senzing/senzing-api-server:1.8.3"
export NEW_IMAGE="myorg/senzing-api-server:1.8.3"

docker build \
    --build-arg BASE_IMAGE=${BASE_IMAGE} \
    --build-arg SENZING_ACCEPT_EULA=${SENZING_ACCEPT_EULA} \
    --tag ${NEW_IMAGE} \
    https://github.com/Senzing/docker-wrap-image-with-senzing-apt.git

2. See error

Step 11/12 : RUN /app/init-container.py initialize-files
 ---> Running in 04bcb084ca7e
2020-05-07 14:14:22,639 senzing-50070297I Enter {"data_dir": "/opt/senzing/data", "debug": false, "enable_db2": false, "enable_mssql": false, "enable_mysql": false, "enable_postgresql": false, "engine_configuration_json": null, "etc_dir": "/etc/opt/senzing", "g2_dir": "/opt/senzing/g2", "gid": 1001, "init_container_sleep": 0, "sleep_time_in_seconds": 0, "start_time": 1588860862.6390128, "subcommand": "initialize-files", "uid": 1001, "update_ini_files": false, "var_dir": "/var/opt/senzing"}
2020-05-07 14:14:22,639 senzing-50070154I /var/opt/senzing/sqlite/G2C.db - Creating file by copying /opt/senzing/g2/resources/templates/G2C.db.template
2020-05-07 14:14:22,640 senzing-50070154I /var/opt/senzing/sqlite/G2C_LIBFEAT.db - Creating file by copying /opt/senzing/g2/resources/templates/G2C.db.template
2020-05-07 14:14:22,641 senzing-50070154I /var/opt/senzing/sqlite/G2C_RES.db - Creating file by copying /opt/senzing/g2/resources/templates/G2C.db.template
2020-05-07 14:14:22,642 senzing-50070154I /etc/opt/senzing/cfgVariant.json - Creating file by copying /opt/senzing/g2/resources/templates/cfgVariant.json.template
Traceback (most recent call last):
  File "/app/init-container.py", line 1670, in <module>
    globals()[subcommand_function_name](args)
  File "/app/init-container.py", line 1544, in do_initialize_files
    copy_files(config)
  File "/app/init-container.py", line 1074, in copy_files
    os.makedirs(os.path.dirname(target_file), exist_ok=True)
  File "/usr/lib/python3.7/os.py", line 221, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: '/etc/opt/senzing'
The command '/bin/sh -c /app/init-container.py initialize-files' returned a non-zero code: 1

**Expected behavior**
I expect the Docker build to complete

**Desktop (please complete the following information):**

- OS: Linux
- Docker: 19.03.8
- Python 3.7 per (RUN apt python3 in container)

**Additional context**
Looks like Python's makdirs is not a stable API, from https://docs.python.org/3/library/os.html#os.makedirs

Changed in version 3.4.1: Before Python 3.4.1, if exist_ok was True and the directory existed, makedirs() would still raise an error if mode did not match the mode of the existing directory. Since this behavior was impossible to implement safely, it was removed in Python 3.4.1. See bpo-21082.

**Comment by docktermj** [comment]
@joshvid All things going well,  I'll try to get this fixed by early next week.   Fingers crossed.

**Comment by docktermj** [comment]
I agree that there is an issue with `os.makedirs(... exist_ok=True)`.  So that will need to be improved. Example:  https://github.com/Senzing/docker-init-container/blob/370607765a22c7d3a4a852d6ee63b1fc0c8964bb/init-container.py#L1074  Among other instances in `init-container.py`.

However, I did try 
```
export BASE_IMAGE="senzing/senzing-api-server:1.8.3"
export NEW_IMAGE="myorg/senzing-api-server:1.8.3"
export SENZING_ACCEPT_EULA=I_ACCEPT_THE_SENZING_EULA

docker build \
  --build-arg BASE_IMAGE=${BASE_IMAGE} \
  --build-arg SENZING_ACCEPT_EULA=${SENZING_ACCEPT_EULA} \
  --tag ${NEW_IMAGE} \
  https://github.com/Senzing/docker-wrap-image-with-senzing-apt.git
```
and it completed successfully.  As did
```
export BASE_IMAGE="senzing/init-container:1.5.3"
export NEW_IMAGE="dockter/init-container:1.5.3"
export SENZING_ACCEPT_EULA=I_ACCEPT_THE_SENZING_EULA

docker build \
    --build-arg BASE_IMAGE=${BASE_IMAGE} \
    --build-arg SENZING_ACCEPT_EULA=${SENZING_ACCEPT_EULA} \
    --tag ${NEW_IMAGE} \
    https://github.com/Senzing/docker-wrap-image-with-senzing-apt.git
```

**Comment by joshvid** [comment]
Thanks for the update, but I'm not sure on next steps or their timing. 

It's a little odd that a container would behave so differently on different hosts. I suspect something deeper in the Docker VM code, perhaps around directory security.

**Comment by docktermj** [comment]
@joshvid Give version 1.5.4 a shot.  If it doesn't work, let me know.

**Comment by joshvid** [comment]
Thanks @docktermj, looks like we moved the ball. That error is resolved, but now it returns the following: 

`Step 11/12 : RUN /app/init-container.py initialize-files
 ---> Running in f9655fa4f90b
2020-05-12 14:02:37,668 senzing-50070297I Enter {"data_dir": "/opt/senzing/data", "debug": false, "enable_db2": false, "enable_mssql": false, "enable_mysql": false, "enable_postgresql": false, "engine_configuration_json": null, "etc_dir": "/etc/opt/senzing", "g2_dir": "/opt/senzing/g2", "gid": 1001, "init_container_sleep": 0, "sleep_time_in_seconds": 0, "start_time": 1589292157.6684148, "subcommand": "initialize-files", "uid": 1001, "update_ini_files": false, "var_dir": "/var/opt/senzing"}
2020-05-12 14:02:37,679 senzing-50070154I /var/opt/senzing/sqlite/G2C.db - Creating file by copying /opt/senzing/g2/resources/templates/G2C.db.template
2020-05-12 14:02:37,684 senzing-50070154I /var/opt/senzing/sqlite/G2C_LIBFEAT.db - Creating file by copying /opt/senzing/g2/resources/templates/G2C.db.template
2020-05-12 14:02:37,686 senzing-50070154I /var/opt/senzing/sqlite/G2C_RES.db - Creating file by copying /opt/senzing/g2/resources/templates/G2C.db.template
2020-05-12 14:02:37,689 senzing-50070154I /etc/opt/senzing/cfgVariant.json - Creating file by copying /opt/senzing/g2/resources/templates/cfgVariant.json.template
Traceback (most recent call last):
  File "/app/init-container.py", line 1674, in <module>
    globals()[subcommand_function_name](args)
  File "/app/init-container.py", line 1548, in do_initialize_files
    copy_files(config)
  File "/app/init-container.py", line 1076, in copy_files
    shutil.copyfile(source_file, target_file)
  File "/usr/lib/python3.7/shutil.py", line 121, in copyfile
    with open(dst, 'wb') as fdst:
NotADirectoryError: [Errno 20] Not a directory: '/etc/opt/senzing/cfgVariant.json'
The command '/bin/sh -c /app/init-container.py initialize-files' returned a non-zero code: 1`

The only change I made in the Dockerfile was senzing/init-container:1.5.3 -> senzing/init-container:1.5.4

**Comment by docktermj** [comment]
@joshvid  

What environment variables and commands are you using to create the error?

I just tried:
```
export BASE_IMAGE="senzing/senzing-api-server:1.8.3"
export NEW_IMAGE="myorg/senzing-api-server:1.8.3"
export SENZING_ACCEPT_EULA=I_ACCEPT_THE_SENZING_EULA

docker build \
  --no-cache \
  --build-arg BASE_IMAGE=${BASE_IMAGE} \
  --build-arg SENZING_ACCEPT_EULA=${SENZING_ACCEPT_EULA} \
  --tag ${NEW_IMAGE} \
  https://github.com/Senzing/docker-wrap-image-with-senzing-apt.git
```
But didn't see a failure.   Are you running different commands, different BASE_IMAGE, or a different Dockerfile?  (I'm trying to figure out why I'm not able to recreate.)

**Note:**  Before running the command, I updated https://github.com/Senzing/docker-wrap-image-with-senzing-apt  [issue-8](https://github.com/Senzing/docker-wrap-image-with-senzing-apt/issues/8) Might that fix the issue you are seeing?

**Comment by joshvid** [comment]
Well, it works now, both with the github-hosted Dockerfile, and with my local version of the Dockerfile. The only difference between the two is that I copy our license file into the image as well: `COPY senzing.lic /etc/opt/senzing`

One of the things which I just did was remove existing Senzing Docker images and the copies. I'm wondering if `docker build` didn't detect some change in one of the image layers when I retried after the 1.5.3->1.5.4 change. But starting from scratch it sorted it out.

Note that I was able to just remove the images: `veracityid/senzing-api-server` and `senzing/senzing-api-server` and it worked. But just to be sure, I removed all of the Senzing related images and ran it from scratch, using my local version of the Dockerfile which also copies over the license file. The build completed this time so it looks like I'm good to go.

I'm still a little perplexed about the whole matter, particularly that it couldn't be duplicated in your env. Even though I'm on a Mac, these containers are all running based on a Linux VM run by Docker. They should work the same everywhere, and the python should be installed w/in the container, which is based on the Senzing image hierarchy. The only root cause I can think of is difference in Linux distributions between what Docker for Mac uses in their VM vs your Linux test env. Host OS shouldn't matter.
